app/controllers/concerns/dynamic_field_names_cacher.rb

```diff
@@ -0,0 +1,21 @@
+module DynamicFieldTitlesCacher
+  CACHE_KEY = "%{identifier}/%{record_type}/column-keys/%{tail}".freeze
+
+  # Takes an ActiveRecord::Relation so we can make one more query off it
+  def collection_field_names(collection, identifier = "projects-#{@current_project.id}")
+    last_updated_record = collection.order(updated_at: :desc).first
+    # Exit early if the collection is empty.
+    return [] unless last_updated_record
+
+    last_updated = last_updated_record.updated_at.to_i
+
+    key_opts = { identifier: identifier, record_type: last_updated_record.class.name }
+
+    Rails.cache.fetch(CACHE_KEY % key_opts.merge(tail: last_updated)) do
+      # Delete any old key before we write the new one
+      Rails.cache.delete_matched(CACHE_KEY % key_opts.merge(tail: '*'))
+
+      collection.map(&:fields).map(&:keys).uniq.flatten
```

Comment:
Another thought, do you really need a new query on L6 (pre changes) to SELECT and ORDER BY?

The collection is already loaded, so just a `map(&:updated_at).sort` may save you a call. I don't think that query is really used.

One issue in Issues#index is that the `collection` is carrying a LEFT OUTER JOIN with it, so I'm thinking reducing the # of queries is probably good (🤷).